include/btree/chunk_ops.h

```diff
@@ -0,0 +1,129 @@
+/*-------------------------------------------------------------------------
+ *
+ * chunk_ops.h
+ *		OrioleDB abstract page chunk access API.
+ *
+ * Copyright (c) 2024, Oriole DB Inc.
+ *
+ * IDENTIFICATION
+ *	  contrib/orioledb/include/btree/chunk_ops.h
+ *
+ *-------------------------------------------------------------------------
+ */
+#ifndef __BTREE_CHUNK_OPS_H__
+#define __BTREE_CHUNK_OPS_H__
+
+#include "btree/btree.h"
+#include "btree/page_contents.h"
+
+typedef enum BTreeChunkOperationType
+{
+	BTreeChunkOperationInsert,
+	BTreeChunkOperationUpdate,
+	BTreeChunkOperationDelete
+} BTreeChunkOperationType;
+
+typedef enum BTreeChunkType
+{
+	BTreeChunkHiKeys,
+	BTreeChunkNonLeaf,
+	BTreeChunkLeaf
+} BTreeChunkType;
+
+typedef struct BTreeChunkBuilderData
+{
+	BTreeDescr *treeDesc;
+
+	Pointer		builderData;
+	Size		builderMaxSize;
+	Size		builderSize;
+
+	BTreeChunkType chunkType;
+} BTreeChunkBuilderData;
+
+typedef struct PartialPageState PartialPageState;
+
+typedef struct BTreeChunkDesc BTreeChunkDesc;
+
+typedef struct BTreeChunkOps
+{
+	/*
+	 * Estimate the new chunk size after the operation.
+	 */
+	int32		(*estimate_change) (BTreeChunkDesc *desc, OffsetNumber itemOffset,
+									BTreeChunkOperationType operation,
+									Pointer tupleHeader, OTuple tuple);
+
+	/*
+	 * Perform the operation.  A caller should take care about the size of the
+	 * chunk.  Extension of the chunk should be performed before "perform_change",
+	 * and shrinking of the chunk should be performed after "perform_change".
+	 */
+	void		(*perform_change) (BTreeChunkDesc *desc, OffsetNumber itemOffset,
+								   BTreeChunkOperationType operation,
+								   Pointer tupleHeader, OTuple tuple);
+
+	/*
+	 * Compare given key to the chunk tuple.
+	 */
+	OBTreeKeyCmp keycmp;
+
+	/*
+	 * Search for the given key in the chunk tuple.  Return false if the chunk
+	 * was changed concurrently.
+	 */
+	bool		(*search) (BTreeChunkDesc *desc, PartialPageState *partial,
+						   Page page, void *key, BTreeKeyType keyType,
+						   OffsetNumber *itemOffset);
+
+	/*
+	 * Read the tuple from the chunk.  Return false if the chunk was changed
+	 * concurrently.
+	 */
+	bool		(*read_tuple) (BTreeChunkDesc *desc, PartialPageState *partial,
```

Comment:
We need another parameter, `bool *needsFree`, which indicates if the returned tuple is a separate copy that needs to be freed. In the current implementation, it should always be set to false. But in the future, we may have implementations that could return a tuple copy without copying a full chunk into a partial state.

Reply:
Added the argument.